src/main/kotlin/com/jetbrains/snakecharm/codeInsight/completion/SmkRulesAndCheckpointsCompletionContributor.kt

```diff
@@ -151,31 +139,38 @@ private class SmkRulesAndCheckpointsObjectsCompletionProvider : CompletionProvid
 
 // the following functions can be used to implement completion for any section/object referring to rule names
 
-private fun <Psi: SmkRuleOrCheckpoint> collectVariantsForElement(
+private fun collectVariantsForElement(
         element: PsiElement,
-        indexKey: StubIndexKey<String, Psi>,
-        sectionClass: Class<Psi>,
-        currentFileDeclarationsFunction: (SmkFile) -> List<Pair<String, Psi>>
-): MutableList<Pair<String, SmkRuleOrCheckpoint>> {
+        isCheckPoint: Boolean
+): List<Pair<String, SmkRuleOrCheckpoint>> {
     val variants = mutableListOf<Pair<String, SmkRuleOrCheckpoint>>()
 
     val module = ModuleUtilCore.findModuleForPsiElement(element)
     if (module != null) {
-        val results = mutableListOf<Psi>()
-        AbstractSmkRuleOrCheckpointType.addVariantFromIndex(indexKey, module, results, sectionClass)
+        val results = if (isCheckPoint) {
+            AbstractSmkRuleOrCheckpointType.
+                    getVariantsFromIndex(SmkCheckpointNameIndex.KEY, module, SmkCheckPoint::class.java)
+        } else {
+            AbstractSmkRuleOrCheckpointType.getVariantsFromIndex(SmkRuleNameIndex.KEY, module, SmkRule::class.java)
+        }
+
         variants.addAll(results
                 .filter { (it as SmkRuleOrCheckpoint).name != null }
                 .map { (it as SmkRuleOrCheckpoint).name!! to it }
         )
     } else {
-        variants.addAll(currentFileDeclarationsFunction.invoke(element.containingFile.originalFile as SmkFile))
+        if (isCheckPoint) {
+            variants.addAll((element.containingFile.originalFile as SmkFile).collectCheckPoints())
+        } else {
+            variants.addAll((element.containingFile.originalFile as SmkFile).collectRules())
+        }
```

Comment:
code duplication here, substing `variants.addAll((element.containingFile.originalFile as SmkFile).`...`)` is quite long and same in lines 163 and 165. (no action required)